File: scripts/configure.py
#!/usr/bin/env python3
"""Resolve a committed profile, upload it, and verify non-secret device status."""
import argparse
import json
import time
from config_profile import ProfileError, add_profile_arguments, load_profile
from serial_device import open_port, wait_ready


def ready_port(name):
    deadline = time.monotonic() + 30
    while True:
        port = None
        try:
            port = open_port(name)
            wait_ready(port, seconds=max(1, deadline - time.monotonic()))
            return port
        except OSError:
            if port is not None:
                port.close()
            if time.monotonic() >= deadline:
                raise ProfileError('Device port unavailable') from None
            time.sleep(0.25)
        except BaseException:
            if port is not None:
                port.close()
            raise


def request(port, command, success, failures=()):
    port.reset_input_buffer()
    port.write((command + '\n').encode('utf-8'))
    deadline = time.monotonic() + 10
    while time.monotonic() < deadline:
        line = port.readline().decode(errors='replace').strip()
        if any(token in line for token in failures):
            raise ProfileError('Device rejected configuration or failed to save it')
        if success in line:
            return line.split(success, 1)[1]
    raise ProfileError('Device acknowledgment not received; inspect device before retrying')


def status(port):
    result = json.loads(request(port, 'config-status', 'device-config '))
    if (not isinstance(result, dict) or type(result.get('policyRevision')) is not int or
            type(result.get('read_only')) is not bool or not isinstance(result.get('rooms'), dict)):
        raise ProfileError('Invalid device configuration status')
    return result


def configure_device(port_name, config, payload):
    try:
        with ready_port(port_name) as port:
            before = status(port)
            request(port, 'config ' + payload, 'CONFIG_SAVED',
                    ('CONFIG_INVALID', 'CONFIG_BUSY', 'CONFIG_SAVE_FAILED'))
        # The saved acknowledgment precedes reboot. Read readiness afresh, then
        # explicitly query status; never accept an unsolicited pre-upload status.
        with ready_port(port_name) as port:
            after = status(port)
        expected_rooms = {
            room: {kind: fields for kind, fields in policies.items() if fields}
            for room, policies in config.get('rooms', {}).items()
        }
        if (after['policyRevision'] != before['policyRevision'] + 1 or
                after['read_only'] != config.get('read_only', True) or
                after['rooms'] != expected_rooms):
            raise ProfileError('Device configuration status does not match submitted profile')
    except ProfileError:
        raise
    except (Exception, SystemExit):
        # Serial exceptions may contain transmitted bytes; never expose them.
        raise ProfileError('Configuration transfer or status verification failed (values omitted)') from None
    print('Configuration saved; revision, read-only mode, and room policies verified.')


def main(argv=None):
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--port', required=True)
    add_profile_arguments(parser)
    args = parser.parse_args(argv)
    try:
        config, payload = load_profile(args.config, args.env_file)
        configure_device(args.port, config, payload)
    except ProfileError as error:
        parser.error(str(error))


if __name__ == '__main__':
    main()

File: scripts/configure.ts
import { isDeepStrictEqual } from "node:util";
import { cli, main, required, stringOption } from "./common.ts";
import { loadProfile, object, ProfileError } from "./config-profile.ts";
import type { Profile } from "./config-profile.ts";
import { configRejectionTokens, readyPort, request } from "./serial-device.ts";
import type { DevicePort } from "./serial-device.ts";

async function status(port: DevicePort) {
  const result: unknown = JSON.parse(
    await request(port, "config-status", "device-config "),
  );
  if (
    !object(result) ||
    !Number.isSafeInteger(result.policyRevision) ||
    typeof result.read_only !== "boolean" ||
    !Number.isInteger(result.sleep_timeout_seconds) ||
    !object(result.rooms) ||
    !object(result.policy)
  )
    throw new Error();
  return result;
}
export async function configureDevice(
  name: string,
  profile: Profile,
  ready = readyPort,
) {
  try {
    const port = await ready(name);
    let before;
    try {
      before = await status(port);
      await request(
        port,
        `config ${profile.payload}`,
        "CONFIG_SAVED",
        configRejectionTokens,
      );
    } finally {
      await port.close();
    }
    // CONFIG_SAVED precedes reboot; require a fresh readiness event and solicited status.
    const rebooted = await ready(name);
    let after;
    try {
      after = await status(rebooted);
    } finally {
      await rebooted.close();
    }
    const rooms = profile.config.rooms ?? {};
    if (!object(rooms)) throw new Error();
    const normalizePolicy = (policies: unknown) => {
      if (!object(policies)) throw new Error();
      return Object.fromEntries(
        Object.entries(policies).filter(
          ([, fields]) => object(fields) && Object.keys(fields).length,
        ),
      );
    };
    const expected = Object.fromEntries(
      Object.entries(rooms).map(([room, policies]) => [
        room,
        normalizePolicy(policies),
      ]),
    );
    if (
      after.policyRevision !== Number(before.policyRevision) + 1 ||
      after.read_only !== (profile.config.read_only ?? true) ||
      after.sleep_timeout_seconds !==
        (profile.config.sleep_timeout_seconds ?? 300) ||
      !isDeepStrictEqual(after.rooms, expected) ||
      !isDeepStrictEqual(
        after.policy,
        normalizePolicy(profile.config.policy ?? {}),
      )
    )
      throw new ProfileError(
        "Device configuration status does not match submitted profile",
      );
  } catch (error) {
    if (error instanceof ProfileError) throw error;
    // Never expose transport errors or device text: either could include credentials.
    throw new ProfileError(
      "Configuration transfer or status verification failed (values omitted); inspect device before retrying",
    );
  }
  console.log(
    "Configuration saved; revision, read-only mode, sleep timeout, and device/room policies verified.",
  );
}
export async function configure(
  argv = process.argv.slice(2),
  send = configureDevice,
) {
  const { values, positionals } = cli(["port", "config", "env-file"], [], argv);
  if (values.help)
    return console.log(
      "node --run configure -- --port PORT [--config PATH] [--env-file PATH]",
    );
  if (positionals.length) throw new Error("Unexpected positional arguments");
  const port = required(values.port, "port");
  const profile = loadProfile(
    stringOption(values.config),
    stringOption(values["env-file"]),
  );
  await send(port, profile);
}
if (import.meta.main) await main(() => configure());
